File: Solutions/CPP/LinearAlgebra/Vector/tests/test_StandardVector.cpp
#include "Vector.h"
#include "gmock/gmock.h"

TEST(standardVector, constructor_double){
    // Test default constructor
    Vector<double> a;
    EXPECT_THAT( a.size(), 0);

    // Test zero vector
    Vector<double> b(3);
    EXPECT_THAT( b.size(), 3);

    // Test values vector
    Vector<double> c({-1, 2.3, 4.5, -4});
    EXPECT_THAT( c.size(), 4);
}

TEST(standardVector, l2norm_double){
    Vector<double> a({-1, 2, 1, -3, 1});
    EXPECT_THAT( a.l2norm(), 4.0 );
}

TEST(standardVector, constructor_int){
    // Test default constructor
    Vector<int> a;
    EXPECT_THAT( a.size(), 0);

    // Test zero vector
    Vector<int> b(3);
    EXPECT_THAT( b.size(), 3);

    // Test values vector
    Vector<int> c({-1, 2, 4, -4});
    EXPECT_THAT( c.size(), 4);
}

TEST(standardVector, l2norm_int){
    Vector<int> a({-1, 2, 1, -3, 1});
    EXPECT_THAT( a.l2norm(), 4 );
}

File: Solutions/Python/LinearAlgebra/Vector/tests/test_standardVector.py
import pytest
from ..Vector import Vector

def test_constructor():
    # Test default constructor
    a = Vector()
    assert a.size == 0
    assert a.values == []

    # Test zero vector
    b = Vector(size=3)
    assert b.size == 3
    assert b.values == pytest.approx([0, 0, 0])

    # Test values vector
    c = Vector(values=[-1, 2.3, 4.5, -4])
    assert c.size == 4
    assert c.values == pytest.approx([-1, 2.3, 4.5, -4])

    # Check error condition
    with pytest.raises(Exception):
        d = Vector(size=3, values=[1, 2, 3])

def test_l2norm():
    a = Vector(values=[-1, 2, 1, -3, 1])
    assert a.l2norm() == pytest.approx(4)